spec/statement.dd

```diff
@@ -808,60 +814,180 @@ $(GNAME OpApplyParameter):
-            import std.stdio;
-
-            int opApply(Dg)(scope Dg dg)
-            if (ParameterTypeTuple!Dg.length == 2) // foreach with 2 parameters
+            int opApply(Body)(scope Body body)
             {
-                writeln(2);
-                return 0;
-            }
-
-            int opApply(Dg)(scope Dg dg)
-            if (ParameterTypeTuple!Dg.length == 3) // foreach with 3 parameters
-            {
-                writeln(3);
+                pragma(msg, Body);
                 return 0;
             }
         }
 
         void main()
         {
-            foreach (int a, int b; S()) { }  // calls first opApply function
-            foreach (int a, int b, float c; S()) { }  // calls second opApply function
+            foreach (int a, int b; S()) { } // int delegate(ref int, ref int) pure nothrow @nogc @safe
+            foreach (bool b, string s; S()) { } // int delegate(ref bool, ref string) pure nothrow @nogc @safe
+            //foreach (x; S()) { } // Error: cannot infer type for `foreach` variable `x`, perhaps set it explicitly
         }
         --------------
         )
 
+$(H4 $(LNAME2 template-instance-op-apply, `opApply` as an Alias to an Explicit Function Template Instance))
+
+    $(P Like any other function, `opApply` and `opApplyReverse` can be aliases to a function or function template.
+        However, special treatment is given to apply functions that are aliases of an appropriate function template instance.)
+
+    $(P In that case, the aliased function template instance is used for overload selection and `foreach` variable type inference,
+        but the template is instantiated again with the actual delegate type of the `foreach` body.
+        This way, apply functions can infer attributes depending on the attributes of the `body` delegate.)
+
+        $(SPEC_RUNNABLE_EXAMPLE_RUN
+        --------------
+        struct A
+        {
+            // Vanilla
+            int opApply(scope int delegate(long) body) => body(42);
+        }
+        struct B
+        {
+            // Template
+            int opApply(Body)(scope Body body) => body(42);
+        }
+        struct C
+        {
+            // Alias to an explicit function template instance
+            alias opApply = opApplyImpl!(int delegate(long));
+            int opApplyImpl(Body)(scope Body body) => body(42);
+        }
+        void main() @nogc nothrow pure @safe
+        {
+            // Error: `@nogc` function `D main` cannot call non-@nogc function `onlineapp.A.opApply`
+            // Error: `pure` function `D main` cannot call impure function `onlineapp.A.opApply`
+            // Error: `@safe` function `D main` cannot call `@system` function `onlineapp.A.opApply`
+            // Error: function `onlineapp.A.opApply` is not `nothrow`
+            // (You might not encounter all of these errors at once.)
+            static assert(!__traits(compiles, () @safe {
```

Comment:
> `@safe` is inherited from `main`.

In fact, it’s not:
```d
void main() @safe
{
    // static assert fails:
    static assert(!__traits(compiles, {
        int x;
        int* p = &x; // something unsafe
    }));
}
```
Without any attribute, the static assert fails.
I noticed that when I tried out, which is why I put it in in the first place.
Because the DAutoTest doesn’t like failing static asserts, I don’t see much of an alternative to leaving it in.